WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -362,23 +460,44 @@ public async Task<ConnectionConfirmationResponse> ConfirmConnectionAsync(Connect
 				var commitAmountZeroCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests);
 				var commitVsizeZeroCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialRequests);
 
-				if (round.Phase == Phase.InputRegistration)
+				using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
 				{
-					alice.SetDeadlineRelativeTo(round.ConnectionConfirmationTimeout);
-					return new(
-						commitAmountZeroCredentialResponse.Commit(),
-						commitVsizeZeroCredentialResponse.Commit());
+					if (round.Phase == Phase.InputRegistration)
+					{
+						alice.SetDeadlineRelativeTo(round.ConnectionConfirmationTimeout);
+
+						return new(
+							commitAmountZeroCredentialResponse.Commit(),
+							commitVsizeZeroCredentialResponse.Commit());
+					}
+					else if (round.Phase == Phase.ConnectionConfirmation)
+					{
+						// Ensure the input can be added to the CoinJoin
+						round.Assert<ConstructionState>().AddInput(alice.Coin);
```

Comment:
would assigning the return value to a `_`  clarify it? There is no difference between `Check` and `Add` conceptually since the round state is immutable, `Add` is accomplished by assigning it back to the round property.

Reply:
I don't think `_` clarifies it much. The fact the we need a comment to explain what that line does is evidence that it is not self explanatory.

Reply:
what about a check and add method on the round itself? this is where that code is headed ultimately, where `Round` would be responsible for its own state. i think making the tx state itself mutable is a mistake for several reasons, and that having code that is slightly unitintuitive in this way is indeed a little confusing, but all things considered strictly better than having yet another piece of mutable state that we need to update in a phased way.

Reply:
specifically, consider that having a `Check` implies that nobody else changes it in the interim, so that requires the MPTS to *also* have synchronization primitives, exactly the same headache that `CredentialIssuer` has WRT supporting concurrency. if it's a pure function then that is entirely `Round`'s responsibility. I feel very strongly about this, I spent so much time just jumping around between files and thinking about how different parts of mutable state interact with each other, but not for this part, so IMO making it immutable already paid off in terms of maintainability.